src/buildings/warehouse/warehouse_building.gd

```diff
@@ -0,0 +1,128 @@
+var specialized_type: EnumAutoload.ResourceType = EnumAutoload.ResourceType.NONE
+
+const SPEC_STORAGE_MULTIPLIER = 4
+const SPEC_PROD_BONUS = 1
+const BASE_PROD_BONUS = 0.2
+
+func _ready():
+	super()
+	data.type = EnumAutoload.BuildingType.STORAGE
+
+
+func apply_upgrades():
+	if EnumAutoload.UpgradeId.CONSTRUCTION_LOGIC_ADV_LOGISTIC in ResourceManager.current_upgrades:
+			get_node("Range").scale = Vector2(1.5, 1.5)
+	# Wait 2 frame to make sure all Area2D changes are setup correctly
+	await get_tree().physics_frame
+	await get_tree().physics_frame
+	check_for_adjacency_multiplier(data.type)
+
+func set_specialisation(_type: EnumAutoload.ResourceType):
+	specialized_type = _type
+	print("Changed")
+	match specialized_type:
+		EnumAutoload.ResourceType.AIR:
+			sprite.self_modulate = Color(1, 0.8, 0.8)
+		EnumAutoload.ResourceType.WATER:
+			sprite.self_modulate = Color(0.8, 0.8, 1)
+		EnumAutoload.ResourceType.FOOD:
+			sprite.self_modulate = Color(0.8, 1, 0.8)
+		EnumAutoload.ResourceType.METAL:
+			sprite.self_modulate = Color(0.7, 0.9, 0.7)
+		EnumAutoload.ResourceType.NONE:
+			sprite.self_modulate = Color(1, 1, 1)
+	# FOrce emit signal so we can update other buildings bonus multiplier
+	BuildingManager.finished_building(EnumAutoload.BuildingType.NONE)
+
+func get_resource_bonus_prod():
+	var resource_data = {"air": 0, "water": 0, "food": 0, "metal": 0}
+	match specialized_type:
+		EnumAutoload.ResourceType.AIR:
+			resource_data.air = SPEC_PROD_BONUS
+		EnumAutoload.ResourceType.WATER:
+			resource_data.water = SPEC_PROD_BONUS
+		EnumAutoload.ResourceType.FOOD:
+			resource_data.food = SPEC_PROD_BONUS
+		EnumAutoload.ResourceType.METAL:
+			resource_data.metal = SPEC_PROD_BONUS
+		EnumAutoload.ResourceType.NONE:
+			resource_data.water = BASE_PROD_BONUS
+			resource_data.food = BASE_PROD_BONUS
+			resource_data.metal = BASE_PROD_BONUS
+			resource_data.air = BASE_PROD_BONUS
+	return resource_data
+
+
+func get_resource_storage_capacity():
+	var resource_data = {"air": 0, "water": 0, "food": 0, "metal": 0}
+	var base_capacity = data.storage_prod
+	match specialized_type:
+		EnumAutoload.ResourceType.AIR:
+			resource_data.air = base_capacity * SPEC_STORAGE_MULTIPLIER
+		EnumAutoload.ResourceType.WATER:
+			resource_data.water = base_capacity * SPEC_STORAGE_MULTIPLIER
+		EnumAutoload.ResourceType.FOOD:
+			resource_data.food = base_capacity * SPEC_STORAGE_MULTIPLIER
+		EnumAutoload.ResourceType.METAL:
+			resource_data.metal = base_capacity * SPEC_STORAGE_MULTIPLIER
+		EnumAutoload.ResourceType.NONE:
+			resource_data.water = base_capacity
+			resource_data.food = base_capacity
+			resource_data.metal = base_capacity
+			resource_data.air = base_capacity
+	return resource_data
+
+
+func enable_improved_preview():
+	if EnumAutoload.UpgradeId.CONSTRUCTION_LOGIC_STOCK_ANALYSIS in ResourceManager.current_upgrades:
+		get_node("Range/RangeSprite").visible = true
+
+func remove_improved_preview():
+	get_node("Range/RangeSprite").visible = false
```

Comment:
Grab this node on init and store it in an @onready var, `get_node` calls within class methods always feels wrong to me - easier to miss mistakes in the string when it's this far down.